### PR TITLE
Hook IDirect3D9Ex (very basic)

### LINKED ISSUE #31: Direct3DCreate9Ex() hook is missing
It's a function available on Windows 7+ similar to `Direct3DCreate9()`, which returns an `IDirect3D9Ex` object. This differs from `IDirect3D9` by a couple of <...>Ex() functions, which also return <...>Ex objects and it goes on like that a little deeper.

A quick fix of returning an error code from `Direct3DCreate9Ex()` works only for some games (like Shovel Knight), others fail to start, so a proper hook is needed.

### PR DESCRIPTION
**Pull request by YaLTeR**
`Direct3DCreate9Ex()` makes an `IDirect3D9Ex`, which is derived from `IDirect3D9` and has an additional `CreateDeviceEx()` method. Similarly, there are `IDirect3DDevice9Ex` and some other `<...>Ex` objects.

This PR concerns initial hooking of `IDirect3D9Ex`.

Currently `IDirect3D9Ex` is hooked with `CreateDeviceEx()` returning an error. This already allows some games to run correctly (like Shovel Knight and, according to my old commit message, SDL games).

### CODE REVIEW

**Comment by YaLTeR** [comment]
The main problem with `<...>Ex` interfaces is that they override some of the parents' virtual functions, which is why a `map` is necessary. I don't like, however, that the way it's done currently the base `MyDirect3D9` has to include `CreateDeviceEx()` which is a method of `MyDirect3D9Ex` exclusively. Perhaps it's possible to design it in such a way that `MyDirect3D9Ex` includes this, and `MyDirect3D9` doesn't. (all of this will span across other `<...>Ex` object hooks)

File: source/hooks/hooks/d3d9hooks.cpp [hunk 1]
```diff
@@ -959,53 +962,185 @@ namespace Hooks
...
+            reinterpret_cast<FARPROC&>(s_vtable_to_original[*reinterpret_cast<void**>(obj)]. ## name), \
+            __FUNCTION__ ": " #name);
+
+    class MyDirect3D9
+    {
+    protected:
+        /*
+         * Map from pointers to vtables to our structs containing the original function pointers.
+         */
+        static std::map<LPVOID, MyDirect3D9> s_vtable_to_original;
```
Comment (Warepire) [inline source/hooks/hooks/d3d9hooks.cpp]: If you make the second type an `LPVOID`, can you move `CreateDeviceEx()` into `MyDirect3D9Ex` then? You should be able to perform `reinterpret_cast<decltype(this)>()` on it to get the correct pointer back.

File: source/tests/d3d9ex/d3d9ex.cpp [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    MSG msg;
+
+    // Main message loop:
+    bool running = true;
+    while (running)
+    {
+        while (PeekMessage(&msg, nullptr, 0, 0, PM_REMOVE))
+        {
+            if (msg.message == WM_QUIT)
+                running = false;
```
Comment (Warepire) [inline source/tests/d3d9ex/d3d9ex.cpp]: Missing {}

**Comment by Warepire** [comment]
Since this lets more games run (for example Shovel Knight), and after a discussion on IRC, I will merge this in its current form.

**Pull request merged**